src/DurableTask.AzureStorage/Storage/QueueMessage.cs

```diff
@@ -21,7 +21,25 @@ class QueueMessage
         public QueueMessage(CloudQueueMessage cloudQueueMessage)
         {
             this.CloudQueueMessage = cloudQueueMessage;
-            this.Message = this.CloudQueueMessage.AsString;
+            try
+            {
+                this.Message = this.CloudQueueMessage.AsString;
+            }
+            catch (FormatException)
+            {
+                // This try-catch block ensures backward compatibility with DTFx.AzureStorage v1.x.
+                // It should only be executed in scenarios requiring this compatibility.
+                try
+                {
+                    // RawString is an internal property of CloudQueueMessage.
+                    System.Reflection.PropertyInfo rawStringProperty = typeof(CloudQueueMessage).GetProperty("RawString", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance);
+                    this.Message = (string)rawStringProperty.GetValue(this.CloudQueueMessage);
+                }
+                catch (Exception) // If failed to read the RawString property, throw an exception.
+                {
+                    throw new InvalidOperationException("Failed to read the cloud queue message.");
+                }
+            }
```

Comment:
some nits on wording. Also, if you're just going to re-throw the caught exception, you don't need to catch it in the first place :)

```suggestion
            catch (FormatException)
            {
                // This try-catch block ensures forwards compatibility with DTFx.AzureStorage v2.x, which does not guarantee base64 encoding of messages (messages not encoded at all).
                // Therefore, if we try to decode those messages as base64, we will have a format exception that will yield a poison message
                // RawString is an internal property of CloudQueueMessage, so we need to obtain it via reflection.
                System.Reflection.PropertyInfo rawStringProperty = typeof(CloudQueueMessage).GetProperty("RawString", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance);
                this.Message = (string)rawStringProperty.GetValue(this.CloudQueueMessage);
            }
```